internal/app/init/pkg/system/services/kubeadm.go

```diff
@@ -50,71 +54,60 @@ func (k *Kubeadm) PreFunc(data *userdata.UserData) (err error) {
 		return err
 	}
 
-	if data.Services.Kubeadm.IsBootstrap() {
-		// Write out all certs we've been provided
-		certs := []struct {
-			Cert     *x509.PEMEncodedCertificateAndKey
-			CertPath string
-			KeyPath  string
-		}{
-			{
-				Cert:     data.Security.Kubernetes.CA,
-				CertPath: constants.KubeadmCACert,
-				KeyPath:  constants.KubeadmCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.SA,
-				CertPath: constants.KubeadmSACert,
-				KeyPath:  constants.KubeadmSAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.FrontProxy,
-				CertPath: constants.KubeadmFrontProxyCACert,
-				KeyPath:  constants.KubeadmFrontProxyCAKey,
-			},
-			{
-				Cert:     data.Security.Kubernetes.Etcd,
-				CertPath: constants.KubeadmEtcdCACert,
-				KeyPath:  constants.KubeadmEtcdCAKey,
-			},
-		}
-
-		for _, cert := range certs {
-			if cert.Cert == nil {
-				continue
-			}
-			if err = writeKubeadmPKIFiles(cert.Cert, cert.CertPath, cert.KeyPath); err != nil {
-				return err
-			}
-		}
-	} else if data.Services.Kubeadm.IsControlPlane() {
-		if data.Services.Trustd == nil || data.Services.Trustd.BootstrapNode == "" {
-			return nil
-		}
+	// Write out all certs we've been provided
+	if err = writeKubeadmPKIFiles(data); err != nil {
+		return err
+	}
 
-		creds, err := basic.NewCredentials(data.Services.Trustd)
-		if err != nil {
+	// Run kubeadm init phase certs all. This should fill in whatever gaps
+	// we have in the provided certs.
+	if data.Services.Kubeadm.IsBootstrap() {
+		cmd := exec.Command(
+			"kubeadm",
```

Comment:
isn't kubeadm in container? 

probably dumb idea, but we could have had 'kubeadm-bootstrap' service injected for bootstrap which runs before `kubeadm`?

Reply:
`kubeadm` is a binary on the host that we mount into a container when it does the `init` or `join`. We do that because the stock `hyperkube` container doesn't have `kubeadm in it and `kubeadm` has some dependencies on things in `hyperkube`. In the case of certs, there are no dependencies so we can run it directly on the host. Having it be a service is kinda nice for observability reasons (`osctl services`). I'm good either way here TBH. A service has more benefits but happy to do in a follow-up.

Reply:
ah, my bad, got it. i was thinking whether we should unpack hyperkube and friends before we launch kubelet or kubeadm or both (unrelated to this PR, just working on the services startup)

Reply:
Yeah there's definitely a use case for breaking up kubeadm into a bootstrap-like phase. While doing this work, it does start to feel like perhaps we should break up each service into an individual package, but I'd imagine there's some more thought that would need to go into something like that.